File: python/src/problems201_300/implement_queue_using_stacks.py
class ImplementQueueUsingStacks:
    def __init__(self):
        self.pushStack: list[int] = []
        self.popStack: list[int] = []

    def push(self, x: int) -> None:
        self.pushStack.append(x)

    def pop(self) -> int:
        self._fill_pop_stack()
        return self.popStack.pop()

    def peek(self) -> int:
        self._fill_pop_stack()
        return self.popStack[-1]

    def empty(self) -> bool:
        return not self.pushStack and not self.popStack

    def _fill_pop_stack(self) -> None:
        if self.popStack:
            return
        while self.pushStack:
            self.popStack.append(self.pushStack.pop())

File: typescript/src/problems201_300/implement-queue-using-stacks.ts
export default class ImplementQueueUsingStacks {
  pushStack: number[];
  popStack: number[];

  constructor() {
    this.pushStack = [];
    this.popStack = [];
  }

  push(x: number) {
    this.pushStack.push(x);
  }

  pop() {
    this._fillPopStack();
    return this.popStack.pop();
  }

  peek() {
    this._fillPopStack();
    return this.popStack[this.popStack.length - 1];
  }

  empty() {
    return this.pushStack.length === 0 && this.popStack.length === 0;
  }

  _fillPopStack() {
    if (this.popStack.length !== 0) {
      return;
    }
    while (this.pushStack.length !== 0) {
      this.popStack.push(this.pushStack.pop());
    }
  }
}
